### PR TITLE
ZEPPELIN-3105 Notebook not running via REST API after Zeppelin Restart

### PR DESCRIPTION
**Pull request by weand**
### What is this PR for?
Fixes regression introduced with [ZEPPELIN-3013](https://issues.apache.org/jira/browse/ZEPPELIN-3013). Executing Notebook and
Paragraph via Rest API was not possible anymore, because some parsing
was introduced paragraph.setText(...). The setter was not invoked after
deserialization, thus transient variables paragraph.intpText and paragraph.scriptText were
not set.

This PR fixes that by invoking paragraph.parseText() after deserialization.

Also a test was added, and AbstractTestRestApi.shutdown() method was
overloaded to support the use case of not deleting the interpreter.json
during server restart within one test case.

### What type of PR is it?
[Bug Fix]

### Todos
-

### What is the Jira issue?
https://issues.apache.org/jira/browse/ZEPPELIN-3105

### How should this be tested?
Refer to the test case added,
org.apache.zeppelin.rest.NotebookRestApiTest.testRunWithServerRestart()

### Questions:
* Does the licenses files need update? - no
* Is there breaking changes for older versions? - no
* Does this needs documentation? - no

### CODE REVIEW

File: zeppelin-zengine/src/main/java/org/apache/zeppelin/notebook/Paragraph.java [hunk 1]
```diff
@@ -830,4 +839,9 @@ public static Paragraph fromJson(String json) {
     return Note.getGson().fromJson(json, Paragraph.class);
   }
 
+  @Override
+  public void postJsonDeser() {
```
Comment (zjffdu) [inline zeppelin-zengine/src/main/java/org/apache/zeppelin/notebook/Paragraph.java]: When this method is called ?

Comment (weand) [reply zeppelin-zengine/src/main/java/org/apache/zeppelin/notebook/Paragraph.java]: Missed commiting of zeppelin-zengine/src/main/java/org/apache/zeppelin/notebook/Note.java. Have a look there.

Comment (zjffdu) [reply zeppelin-zengine/src/main/java/org/apache/zeppelin/notebook/Paragraph.java]: we can call `parseText` in  method `fromJson` which is much easier. `fromJson` is the method which is called when derealization.

Comment (weand) [reply zeppelin-zengine/src/main/java/org/apache/zeppelin/notebook/Paragraph.java]: adjusted code according to your comment. Can you review again please.

File: zeppelin-server/src/test/java/org/apache/zeppelin/rest/AbstractTestRestApi.java [hunk 2]
```diff
@@ -315,6 +315,10 @@ private static boolean isActiveSparkHome(File dir) {
   }
 
   protected static void shutDown() throws Exception {
+    shutDown(true);
+  }
+
+  protected static void shutDown(final boolean deleteConfDir) throws Exception {
```
Comment (zjffdu) [inline zeppelin-server/src/test/java/org/apache/zeppelin/rest/AbstractTestRestApi.java]: What's the purpose for this change ? whether delete conf is controlled by `ZeppelinConfiguration`, it seems not necessary to be called by user explicitly.

File: zeppelin-zengine/src/main/java/org/apache/zeppelin/notebook/Note.java [hunk 3]
```diff
@@ -20,21 +20,31 @@
 import static java.lang.String.format;
 
 import java.io.IOException;
-import java.util.*;
+import java.util.Date;
```
Comment (zjffdu) [inline zeppelin-zengine/src/main/java/org/apache/zeppelin/notebook/Note.java]: Zeppelin need to have a unified code style (whether import * or import each separately), would be appreciated someone can do that (adding checkstyle)

**Comment by zjffdu** [comment]
Thanks @weand mostly LGTM, just a minor issue

File: zeppelin-server/src/test/java/org/apache/zeppelin/rest/AbstractTestRestApi.java (hunk 2 above)
Comment (weand) [reply zeppelin-server/src/test/java/org/apache/zeppelin/rest/AbstractTestRestApi.java]: I did not see another proper chance to avoid deletion of intepreter.json file from the newly added test case. the test case needs to restart the server, which would delete the interpreter bindings without that change.

File: zeppelin-server/src/test/java/org/apache/zeppelin/rest/AbstractTestRestApi.java [hunk 4]
```diff
@@ -352,7 +356,7 @@ protected static void shutDown() throws Exception {
             .clearProperty(ZeppelinConfiguration.ConfVars.ZEPPELIN_ANONYMOUS_ALLOWED.getVarName());
       }
 
-      if (!ZeppelinServer.notebook.getConf().isRecoveryEnabled()) {
+      if (deleteConfDir && !ZeppelinServer.notebook.getConf().isRecoveryEnabled()) {
```
Comment (zjffdu) [inline zeppelin-server/src/test/java/org/apache/zeppelin/rest/AbstractTestRestApi.java]: I see

File: zeppelin-server/src/test/java/org/apache/zeppelin/rest/AbstractTestRestApi.java (hunk 2 above)
Comment (zjffdu) [reply zeppelin-server/src/test/java/org/apache/zeppelin/rest/AbstractTestRestApi.java]: I see

**Comment by zjffdu** [comment]
@weand CI still fails, but very probably it is flaky test, but could you retrigger the CI ? I'd like make sure these failed test is not caused by this PR before merging.

**Comment by weand** [comment]
Tests run: 3, Failures: 0, Errors: 3, Skipped: 0, Time elapsed: 164.418 sec <<< FAILURE! - in org.apache.zeppelin.integration.InterpreterModeActionsIT

seems to be a flaky test, see log [here](https://travis-ci.org/weand/zeppelin/jobs/318812306)

**Comment by felixcheung** [comment]
you can close and reopen this PR to trigger CI

**Comment by weand** [comment]
@zjffdu can u confirm it's a flaky test?

**Comment by zjffdu** [comment]
What's the result of your rerun ? I don't see this test fails frequently in other PRs.

**Comment by weand** [comment]
close and reopen PR does NOT trigger the CI.

now merged latest master again, and awaiting test results.

**Comment by weand** [comment]
This time it failed here:
org.apache.zeppelin.integration.InterpreterModeActionsIT.testPerUserIsolatedAction(InterpreterModeActionsIT.java:729)

the last time it failed at another line - 660:
org.apache.zeppelin.integration.InterpreterModeActionsIT.testPerUserIsolatedAction(InterpreterModeActionsIT.java:660)

this seems quite flaky. Can not see, why the changes here break this test. Do you have any idea?

**Comment by zjffdu** [comment]
If you see this failed test consistently, I would suspect it may due to this PR. Because I didn't see this failed test consistently in other PRs.

**Comment by weand** [comment]
Don't get the Selenium tests running properly.

Instead of the error from Travis CI, I face other errors, that number of python processes is wrong.

```
21:05:52,088  INFO org.apache.zeppelin.CommandExecutor:46 - Sending command "bash -c ps aux | grep 'zeppelin_ipython' | grep -v 'grep' | wc -l" to localhost
21:05:52,290  INFO org.apache.zeppelin.CommandExecutor:46 - Sending command "bash -c ps aux | grep 'zeppelin/interpreter/python/*' | sed -E '/grep|local-repo/d' | wc -l" to localhost
21:05:56,226  INFO org.apache.zeppelin.CommandExecutor:46 - Sending command "bash -c ../bin/zeppelin-daemon.sh restart" to localhost
Tests run: 11, Failures: 11, Errors: 0, Skipped: 0, Time elapsed: 471.138 sec <<< FAILURE! - in org.apache.zeppelin.integration.InterpreterModeActionsIT
testPerUserIsolatedAction(org.apache.zeppelin.integration.InterpreterModeActionsIT)  Time elapsed: 186.266 sec  <<< FAILURE!
java.lang.AssertionError: The number of python process is
Expected: "1"
     but: was "0"
	at org.hamcrest
```

I already have ipython + jupyter installed. Also Firefox 31 with Auto Update disabled.

When starting zeppelin and a python notebook (created by the tests) manually, I get proper results inside Web GUI. I also see various zeppelin processes via ps aux. But the grep pipes used in the test doesn't cover them.

Could anyone please assist, who can run the test properly. @soralee ?

**Comment by weand** [comment]
@zjffdu As this fixes a regression from [ZEPPELIN-3013](https://issues.apache.org/jira/browse/ZEPPELIN-3013) would you mind helping here out?

**Comment by zjffdu** [comment]
Sorry @weand there's one critical issue in travis build, will come back after fixing that issue. Feel free to ping me in case I forget.

**Comment by zjffdu** [comment]
@weand Travis build issue has been fixed, could you retrigger the build of this PR ?

**Comment by zjffdu** [comment]
Thanks @weand merge it
